### PR TITLE
feat(nargo)!: define preprocessed artifacts for programs/contracts

### PR DESCRIPTION
**Pull request by TomAFrench**
# Related issue(s)

Related to #1102

# Description

## Summary of changes

This PR defines generic preprocessed artifacts for Noir programs and contracts as defined in #1102 ("Backend-specific Artifact").

Rather than the current `json/pk/vk/checksum` files we now have a single json file which we can use. This means that we don't need to be concerned about the keys and the acir being in sync as they're stored together.

This does mean that we now need to load the proving key when verifying a proof however my preference for solving this is to derive specialized "verifier artifacts" which just contain the information required for verifying the circuit. Nargo would then accept the full or verifier artifact when verifying a circuit but that's OOS for this PR.

Note: `PreprocessedContract` does NOT match the desired spec for A3 contracts, these can be derived from this artifact however. I don't think that Nargo is the right place for performing this translation so the A3 team may need to maintain a small tool to translate between them. Thoughts @kevaundray? 

## Test additions / changes

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

## Documentation needs
- [x] This PR requires documentation updates when merged.

Nargo compilation outputs have been changed. This shouldn't affect any actions which the user has to take but discussion about files created needs to be update.

`nargo preprocess` command has been removed as it's not meaningful until we have backend-agnostic artifacts. The `nargo preprocess` command would then create the artifacts made in this PR.

# Additional context

### CODE REVIEW

**Comment by phated** [comment]
> This does mean that we now need to load the proving key when verifying a proof however my preference for solving this is to derive specialized "verifier artifacts" which just contain the information required for verifying the circuit. Nargo would then accept the full or verifier artifact when verifying a circuit but that's OOS for this PR.

Why would we have specialized artifacts instead of just switching to a serialization protocol that can read individual pieces of data out of the artifact?

**Comment by TomAFrench** [comment]
> Why would we have specialized artifacts instead of just switching to a serialization protocol that can read individual pieces of data out of the artifact?

Both are options. Chosing a serialisation format which allows us to read subsets of the artifact is obviously better when you're storing the full artifact anyway but only need a portion of it for the action you're performing. However, if we know that we're never going to prove against the circuit (e.g. we want to verify proofs on a mobile device but it doesn't have the resources to create proofs) then we can save bandwidth/disk space using specialised artifacts.

**Comment by TomAFrench** [comment]
Removing generic from the title as this causes some confusion in relation to #1102. For clarity these artifacts aren't generic wrt to backends but are "generic" in the sense that they're not tailored to any particular usecase.

File: crates/nargo_cli/src/preprocess.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        proving_key,
+        verification_key,
+    }
+}
+
+pub(crate) fn preprocess_contract(compiled_contract: CompiledContract) -> PreprocessedContract {
+    let backend = crate::backends::ConcreteBackend;
+
+    let preprocessed_contract_functions = vecmap(compiled_contract.functions, |func| {
+        // TODO: currently `func`'s bytecode is already optimized for the backend.
```
Comment (kevaundray) [inline crates/nargo_cli/src/preprocess.rs:36]: This TODO is okay to merge with -- can you create an issue for it?

File: crates/nargo_cli/src/preprocess.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    program::PreprocessedProgram,
+};
+
+// TODO: pull this from backend.
+const BACKEND_IDENTIFIER: &str = "acvm-backend-barretenberg";
+
+pub(crate) fn preprocess_program(compiled_program: CompiledProgram) -> PreprocessedProgram {
+    let backend = crate::backends::ConcreteBackend;
+
+    // TODO: currently `compiled_program`'s bytecode is already optimized for the backend.
```
Comment (kevaundray) [inline crates/nargo_cli/src/preprocess.rs:18]: This is the same TODO from below

File: crates/nargo_cli/src/preprocess.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+use noirc_driver::{CompiledContract, CompiledProgram};
+
+// TODO: migrate to `nargo_cli`
+
+use crate::artifacts::{
+    contract::{PreprocessedContract, PreprocessedContractFunction},
+    program::PreprocessedProgram,
+};
+
+// TODO: pull this from backend.
```
Comment (kevaundray) [inline crates/nargo_cli/src/preprocess.rs:12]: Can this be done before merging?

File: crates/nargo_cli/src/preprocess.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
+use acvm::ProofSystemCompiler;
+use iter_extended::vecmap;
+use noirc_driver::{CompiledContract, CompiledProgram};
+
+// TODO: migrate to `nargo_cli`
```
Comment (kevaundray) [inline crates/nargo_cli/src/preprocess.rs:5]: Can this be done before merging?

File: crates/nargo_cli/src/preprocess.rs (hunk 3 above)
Comment (TomAFrench) [reply crates/nargo_cli/src/preprocess.rs:12]: The backend doesn't currently expose this (and we can't push a new release without solving UP).

File: crates/nargo_cli/src/preprocess.rs (hunk 4 above)
Comment (TomAFrench) [reply crates/nargo_cli/src/preprocess.rs:5]: I'm guessing this should read `nargo-core`. Yeah, I was planning on addressing this in a followup PR which moves a bunch of similar logic over in one go but I can address this now (or reverse the order of the PRs).

File: crates/nargo_cli/src/preprocess.rs (hunk 3 above)
Comment (kevaundray) [reply crates/nargo_cli/src/preprocess.rs:12]: Makes sense -- lets create a tracking issue for this

Comment (TomAFrench) [reply crates/nargo_cli/src/preprocess.rs:12]: https://github.com/noir-lang/acvm/issues/196

**Pull request merged**